### PR TITLE
#777-GanttProjectSort by WBSNum

### LINKED ISSUE #777: [Frontend] - Gantt Projects & Work Packages Sort Order
### Browser(s)

Chrome

### Platform(s)

Desktop

### Version(s)

production (main)

### Observed Behavior

The projects and work packages in the Gantt Chart on the Gantt Page appear by their order in the database.

### Expected Behavior

They should be sorted by their WBSNum.

### Steps to Reproduce

1. Go to the Gantt Page
2. Observe the order of the projects
3. Expand and observe the order of work packages

### Screenshots (as needed)

image

### PR DESCRIPTION
**Pull request by ahnfikd7**
## Changes
Replaced  proj.workPackages.sort((a, b) => a.startDate.getTime() - b.startDate.getTime());
to workPackages.sort((a, b) => a.wbsNum.workPackageNumber - b.wbsNum.workPackageNumber);
_Explanation of changes goes here_
Hopefully due to this change, workpackages will sort due to wbsNum.

Update: Changed the sorting technique in GanttPageWrapper

## To Do

- [x] Unit Testing

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [ ] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #777

### CODE REVIEW

**Review by Ari-S-123 — CHANGES_REQUESTED** [review]
You have to sort the projects and work packages on the Gantt Page, probably in `GanttPageWrapper.tsx`, not in the project detail page. This one is a bit more complicated than I thought at first so if you need more help, you can ask for help in slack. Also you have first set up the testing conditions by changing some of the WBSNums for the seed data in `projects.seed.ts`. Change the first seed project to 1.10.0, the second one to 1.9.0,  and the third one to 2.8.0. You won't be committing these but you need these for reproducing this bug on localhost.

**Comment by ahnfikd7** [comment]
Do you mean changing the projectNumber or workPackageNumber in projects.seed.ts in each of the dbSeedProject? Because I don't see anything named wbsNum here.  Also, I can't find anything named GanttPageWrapper.tsx in my local repo, the Gantt tab on the side doesn't show up in my localhost.

**Comment by Ari-S-123** [comment]
Just change the car number, project number, and work package according to the examples I gave for testing locally.

> Also, I can't find anything named GanttPageWrapper.tsx in my local repo, the Gantt tab on the side doesn't show up in my localhost.

Update your local branch

**Comment by ahnfikd7** [comment]
Let me know if I need to make any more changes.

File: src/frontend/src/pages/GanttPage/GanttPageWrapper.tsx [hunk 1]
```diff
@@ -75,6 +87,18 @@ const GanttPageWrapper: FC = () => {
 
   useEffect(() => {
     const transformProjectToGanttObject = (project: Project): Task => {
+      project.workPackages.sort((a, b) => {
```
Comment (z-roth) [inline src/frontend/src/pages/GanttPage/GanttPageWrapper.tsx:90]: Looks great - only thing is that there's a bit of code duplication here. Can we abstract this to one function that we can use in both cases?

File: src/frontend/src/pages/GanttPage/GanttPageWrapper.tsx [hunk 2]
```diff
@@ -75,6 +87,18 @@ const GanttPageWrapper: FC = () => {
...
+        const aWbsNum = a.wbsNum as WbsNumber;
```
Comment (z-roth) [inline src/frontend/src/pages/GanttPage/GanttPageWrapper.tsx:91]: Do we need to cast these?

**Review by z-roth — CHANGES_REQUESTED** [review]

**Comment by ahnfikd7** [comment]
I fixed the duplication and casting as well. I'm not sure why the prettier check is failing in change-request-types.ts. I didn't even touch that file. I tried to do Alt+Shift+F formatting,  but that doesn't fix it either. This is happening after I did git pull, before pushing my latest code.

**Comment by ahnfikd7** [comment]
Fixed!

**Review by Ari-S-123 — CHANGES_REQUESTED** [review]
This ticket shouldn't need any changes to `package.json` or `yarn.lock`

**Comment by ahnfikd7** [comment]
> This ticket shouldn't need any changes to `package.json` or `yarn.lock`

I didn't even touch those files, what might be the issue?

**Comment by ahnfikd7** [comment]
I copy pasted the yarn.lock and package.json files from the develop branch into my local machine and then ran yarn install. That might fix the issue with changes of yarn.lock and package.json. However, the change-request-types.ts prettier check failure is back, I don't know how to fix that

File: src/shared/src/types/change-request-types.ts [hunk 3]
```diff
@@ -29,7 +29,7 @@ export const ChangeRequestType = {
   Activation: 'ACTIVATION'
 } as const;
 // eslint-disable-next-line @typescript-eslint/no-redeclare
-export type ChangeRequestType = typeof ChangeRequestType[keyof typeof ChangeRequestType];
+export type ChangeRequestType = (typeof ChangeRequestType)[keyof typeof ChangeRequestType];
```
Comment (z-roth) [inline src/shared/src/types/change-request-types.ts:32]: Is there a reason for this change?

**Review by z-roth — CHANGES_REQUESTED** [review]

File: src/shared/src/types/change-request-types.ts (hunk 3 above)
Comment (ahnfikd7) [reply src/shared/src/types/change-request-types.ts:32]: If I type the formatting shortcut Shift + Alt + F, then it does that automatically

**Review by z-roth — APPROVED** [review]

**Comment by z-roth** [comment]
Looks good to me - nice work!

**Review by Ari-S-123 — CHANGES_REQUESTED** [review]
Doesn't appear to work, this pic shows the order they are in the database. They should be sorted by their WBS instead.
Screenshot_20230216_101102

**Comment by Ari-S-123** [comment]
**Database Order:**

Projects:
1.10.0
1.9.0
2.8.0
0.12.0

Work Packages:
1.10.2
1.10.1

**Display Order:**
Screenshot_20230216_103957

**Review by Ari-S-123 — APPROVED** [review]

**Review by Ari-S-123 — CHANGES_REQUESTED** [review]
just noticed another bug

**Comment by Ari-S-123** [comment]
**Database Order:**

Projects:
1) 2.8.0
2) 1.10.0
3) 1.9.0
4) 0.12.0
5) 1.25.3

Work Packages:
1) 2.8.2
2) 2.8.1

Screenshot_20230216_113721

**Review by Ari-S-123 — DISMISSED** [review]

File: src/frontend/src/pages/GanttPage/GanttPageWrapper.tsx [hunk 4]
```diff
@@ -112,9 +132,21 @@ const GanttPageWrapper: FC = () => {
...
-      const projTasks = filteredProjects.map(transformProjectToGanttObject);
-      const workPackages = filteredProjects.flatMap((p) => p.workPackages);
-      const wpTasks = workPackages.map((wp) => transformWPToGanttObject(wp, filteredProjects));
+      const sortedProjects = filteredProjects.sort(sortWbs);
+      const indexedProjects = sortedProjects.map((p) => {
+        const displayProject = p as GanttDisplayProject;
+        displayProject.displayOrder = sortedProjects.indexOf(displayProject) + 1;
+        return displayProject;
+      });
+      const projTasks = indexedProjects.map(transformProjectToGanttObject);
```
Comment (anthonybernardi) [inline src/frontend/src/pages/GanttPage/GanttPageWrapper.tsx:141]: projectTasks

**Review by anthonybernardi — CHANGES_REQUESTED** [review]
let's put the sortWbs function in a file `gantt.utils.ts` in the GanttPage folder and then write a few tests for it

also the logic is incredibly convoluted for all the list transformations we're doing. I'm sure it doesn't need to be this complicated but if it does there should at least be comments explaining everything

**Comment by anthonybernardi** [comment]
image
works with prod data

**Review by anthonybernardi — APPROVED** [review]
noice

**Pull request merged**